modules/dnn/src/cuda4dnn/csl/tensor.hpp

```diff
@@ -0,0 +1,807 @@
+                0.0, result.get(), dest_nc);
+        }
+
+        /** @brief performs matrix-addition
+         *
+         * Pre-conditions:
+         * - \p A and \p B must meet the mathematical requirements for matrix addition
+         * - \p result must be large enough to hold the result
+         *
+         * Exception Gaurantee: Basic
+         */
+        template <class T> inline
+        void add(const cublas::Handle& handle, TensorSpan<T> result, TensorView<T> A, TensorView<T> B) {
+            /* matrix operations can be performed only on rank two tensors */
+            using tensor_utils::get_effective_rank;
+            CV_Assert(get_effective_rank(A) == 2 &&
+                      get_effective_rank(B) == 2 &&
+                      get_effective_rank(result) == 2);
+
+            /* check dimension requirements for matrix addition */
+            CV_Assert(tensor_utils::is_same_shape(A, B));
+            CV_Assert(tensor_utils::is_same_shape(A, result));
+
+            const auto dest_nr = result.get_axis_size(-2);
+            const auto dest_nc = result.get_axis_size(-1);
+            const auto A_nr = A.get_axis_size(-2);
+            const auto B_nr = B.get_axis_size(-2);
+
+            cublas::geam<T>(handle,
+                false, false,
+                dest_nr, dest_nc,
+                1.0, A.get(), A_nr,
+                1.0, B.get(), B_nr,
+                result.get(), dest_nr);
+        }
+
+        template <class T> inline
+        void abs(const Stream& stream, TensorSpan<T> dest, TensorView<T> src) {
+            CV_Assert(tensor_utils::is_same_shape(dest, src));
+            kernels::abs(stream, span<T>(dest.get(), dest.size()), view<T>(src.get(), src.size()));
+        }
+
+        template <class T> inline
+        void bnll(const Stream& stream, TensorSpan<T> dest, TensorView<T> src) {
+            CV_Assert(tensor_utils::is_same_shape(dest, src));
+            kernels::bnll(stream, span<T>(dest.get(), dest.size()), view<T>(src.get(), src.size()));
+        }
+
+        template <class T> inline
+        void relu(Stream stream, TensorSpan<T> dest, TensorView<T> src, T slope = 0) {
+            CV_Assert(tensor_utils::is_same_shape(dest, src));
+            kernels::relu(stream, span<T>(dest.get(), dest.size()), view<T>(src.get(), src.size()), slope);
+        }
+
+        template <class T> inline
+        void clipped_relu(const Stream& stream, TensorSpan<T> dest, TensorView<T> src, T max, T min = 0) {
+            CV_Assert(tensor_utils::is_same_shape(dest, src));
+            kernels::clipped_relu(stream, span<T>(dest.get(), dest.size()), view<T>(src.get(), src.size()), max, min);
+        }
+
+        template <class T> inline
+        void channelwise_relu(const Stream& stream, TensorSpan<T> dest, TensorView<T> src, TensorView<T> slope) {
+            CV_Assert(tensor_utils::is_same_shape(dest, src));
+            CV_Assert(src.get_axis_size(1) == slope.size());
+            CV_Assert(0); // TODO
```

Comment:
I can execute the ReLU kernel for each channel separately or I could write a single kernel to apply channel-wise ReLU activation.

Single kernel launch:
- requires a GPU tensor to store the slopes
- eliminates multiple kernel launches

Multiple kernels:
- GPU slope tensor isn't required
- multiple kernel launches might add a performance penalty (small?)

I don't know if it'll have any impact in practical applications as applying ReLU is way faster compared to convolution or even gemm.

Reply:
You definitely want each kernel to process a whole batch. There isn't a lot of overhead for a kernel launch, but it is still non-trivial. More importantly, kernels launched on the same stream are not run in parallel. So you will get really awful GPU utilization if you launch multiple kernels, each processing some fraction of a batch.  The main reason people use this batching in deep learning is so that GPUs can process each batch step in one big parallel kernel. So it's super important that each DNN layer launches a number of kernels that isn't a function of the number of batch elements or number of channels or anything like that.  For the relu layer, as an example, you should launch exactly one kernel that does the whole computation in one shot.